Implement resolve code for ConstantDynamic entries

fengxue-IS opened the pull request:
- added ConstantDynamic case to ldc, ldc_w, ldc2_w bytecode
- added native resolve helper for ConstantDynamic
- new MethodHandle resolve method for Java side
- implement mutex on ramslot to ensure single resolve is allowed

Close: #1272

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
depend: #1426

DanHeidinga commented:
Once  #1426 is merged (should be later today when the builds finish), can you rebase this to remove the changes from  #1426?

fengxue-IS commented:
sure, I ll update the PR after #1426 is merged

fengxue-IS commented:
branch rebase on master

DanHeidinga commented on runtime/vm/j9vm.tdf:
```diff
@@ -561,6 +561,9 @@ TraceExit=Trc_VM_jitFillOSRBuffer_Exit Overhead=1 Level=3 Template="jitFillOSRBu
 TraceEntry=Trc_VM_sendResolveMethodHandle_Entry Overhead=1 Level=2 Template="sendResolveMethodHandle"
 TraceExit=Trc_VM_sendResolveMethodHandle_Exit Overhead=1 Level=2 Template="sendResolveMethodHandle"
 
+TraceEntry=Trc_VM_sendResolveConstantDynamic_Entry Overhead=1 Level=2 Template="sendResolveConstantDynamic"
```
Tracepoints can only be added to the end of the file

fengxue-IS replied:
fixed

DanHeidinga commented on runtime/vm/vmthinit.c:
```diff
@@ -80,6 +80,8 @@ UDATA initializeVMThreading(J9JavaVM *vm)
 
 		omrthread_monitor_init_with_name(&vm->vmRuntimeStateListener.runtimeStateListenerMutex, 0, "VM state notification mutex") ||
 
+		omrthread_monitor_init_with_name(&vm->constantDynamicMutex, 0, "Lock for Constant Dynamic resolve store") ||
```
This is a single lock shared by all threads and all constantdynamics.  We should be able to resolve without a lock

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,93 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	UDATA resolved = 0;
+
+	/* Check if already resolved */
+	omrthread_monitor_enter(vmThread->javaVM->constantDynamicMutex);
```
A memory barrier should be sufficient here to see if this has been resolved already or not.

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,93 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		value = (j9object_t)vmThread->returnValue;
+
+		/* lock write access to constantDynamic ram slots */
+		omrthread_monitor_enter(vmThread->javaVM->constantDynamicMutex);
```
Similar concern here.  We need to spend some time thinking about the ordering, memory barriers, and CAS to handle setting and reading these fields without a lock

tajila commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -828,6 +849,150 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+				case 14:
+					cpEntry = getCPMethodHandleAt(internalRamClass, index);
+					break;
+				default:
```
also need a `case 17:` for constant dynamic. Right now this will be a 
```
//TODO
break;
```

until https://github.com/eclipse/openj9/pull/1649 is complete

`resolveInvokeDynamic` will have to be updated as well

tajila commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -828,6 +849,150 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+			default:
+				result = (Object) bsm.invokeWithArguments(staticArgs);
+				break;
+			}
```
need to do the identity check here

fengxue-IS replied:
fixed

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	J9Class *ramClass = ramCP->ramClass;
+	J9ROMClass *romClass = ramClass->romClass;
+	J9ROMConstantDynamicRef *romConstantRef = (J9ROMConstantDynamicRef*)(J9_ROM_CP_FROM_CP(ramCP) + cpIndex);
+	J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
```
you can declare callSiteData, bsmIndices, bsmData, bsmIndex and nameAndSig later since they are only being used in the inner scope

tajila replied:
same with romClass and romConstantRef

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	U_32 bsmIndex = romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift;
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
```
you can also declare that later

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	J9Class *throwable = internalFindKnownClass(vmThread, J9VMCONSTANTPOOL_JAVALANGTHROWABLE, J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
```
j.l.Throwable should be loaded  before any exception object is created. So by the time you need it you should be able to get it directly from the JCl constant pool.

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	J9Class *throwable = internalFindKnownClass(vmThread, J9VMCONSTANTPOOL_JAVALANGTHROWABLE, J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
+	UDATA resolved = 0;
```
you can make this a BOOLEAN since you are only using it for binary checks

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			/* Resolved with exception */
+			VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+			resolved = 1;
+		} else if (ramCPEntry->exception == vmThread->javaVM->voidReflectClass->classObject) {
```
you are using javaVM multiple times, you can cache a copy in a local variable

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		/* Invoke BootStrap method*/
+		sendResolveConstantDynamic(vmThread, ramCP, cpIndex, nameAndSig, bsmData);
+		value = (j9object_t)vmThread->returnValue;
+
```
better to move the `(ramCPEntry->exception != vmThread->threadObject)` check here, no point even running the rest of the code if someone else has resolved it

probably need read barrier before the check

fengxue-IS replied:
fixed

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			resolved = 1;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			// yield(vmThread);
```
VM_AtomicSupport::yieldCPU();

tajila replied:
Actually, yield my not be sufficient to ensure that the resolving thread gets precedence.

You can add a monitor (ie. condyMonitor) to each j9vmthread. Then the thread can wait on `((J9VMThread *)J9VMJAVALANGTHREAD_THREADREF(ramCPEntry->exception))->condyMonitor`.

And the code that writes value can notify condyMonitor

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			}
+		}
+
+		if (ramCPEntry->exception == vmThread->threadObject) {
```
in order to reach this point, doesn't this check always have to be true?

fengxue-IS replied:
in the case of nested condy, the inner most call could resolve before reaching here

tajila commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		}
+
+		if (ramCPEntry->exception == vmThread->threadObject) {
+			ramCPEntry->exception = exceptionObject;
```
might need a write barrier before this. I'm not sure how much reordering the compiler will do, but if it is possible to reoder this before the write to `ramCPEntry->value`, then you definitely need a write barrier.

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1839,6 +1839,101 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			resolved = TRUE;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			VM_AtomicSupport::yieldCPU();
```
@DanHeidinga @gacholio In the case that another thread is currently resolving the entry, would thread yield be sufficient to wait until the resolving thread finish?

gacholio replied:
No, yield is essentially a no-op - it guarantees nothing.

fengxue-IS replied:
sry, I meant if yield is used here, would it be sufficient (as in suspending the thread for short period) so it will not be consistently spinning in the `goto retry;` loop as other thread is resolving the condy entry. (so we don't have this thread wasting resources while not doing any actual work, a mutex would be ideal but too expensive?)

DanHeidinga commented:
@fengxue-IS Tobi, GAC, and I sketched out a possible resolve algorithm.  We should sit down together and go over it to get your input and then update this PR.

tajila commented:
We might want to think about making a common helper for `MethodHandle.resolve[Constant|Invoke]Dynamic` and `sendResolve[Invoke|Constant}Dynamic` in a follow on PR. Rest lgtm.

@gacholio please take a look

DanHeidinga commented:
@gacholio Can you review this?

gacholio commented:
I'd like to see @DanHeidinga approve first, and the conflicts resolved.

fengxue-IS commented:
Rebased on master branch, conflicts resolved. All the review comments have been addressed.
Tests pending

gacholio commented:
I'll review this today or tomorrow.  @DanHeidinga can approve when he gets a chance.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = ramCPEntry->value;
+	J9JavaVM *vm = vmThread->javaVM;
+	BOOLEAN resolved = FALSE;
```
This is a cpp file, so please use bool.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			resolved = TRUE;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			omrthread_monitor_wait(vm->constantDynamicMutex);
```
I see no indication the monitor is ever entered or exitted.

gacholio commented on runtime/vm/callin.cpp:
```diff
@@ -941,6 +941,51 @@ sendForGenericInvoke(J9VMThread *currentThread, j9object_t methodHandle, j9objec
+			if (NULL != sigString) {
+				/* Run the method */
+				/* skip one slot because we are passing a long */
+				currentThread->sp -= 2;
```
Pleaase move this after the long comment so it's beside the store.

fengxue-IS commented:
added `omrthread_monitor_enter/exit()` around `wait` and `notify_all` operation
fixed the comment position as suggested and changed `BOOLEAN` to `bool`

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,106 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		ramCPEntry->exception = exceptionObject;
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		omrthread_monitor_notify_all(vm->constantDynamicMutex);
+		omrthread_monitor_exit(vm->constantDynamicMutex);
```
This is not a legal use of wait/notify.  Let's discuss tomorrow.

gacholio replied:
It would be better to stash the two written values in locals, and update the ramCPEntry while holding the monitor.

fengxue-IS replied:
I know it is a better practice to update the ramCPEntry within the monitor, but not sure how much overhead this would add in a multi-thread case when parallel resolve takes place

gacholio replied:
None at all - it's the same code (modulo a few local variable accesses), just shuffled around.

fengxue-IS replied:
I've made the change to update the cp entry values inside the monitor scope, we can change this in the future if this becomes a performance issue

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,106 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			/* Another thread is currently trying to resolve the entry */
+			omrthread_monitor_enter(vm->constantDynamicMutex);
+			omrthread_monitor_wait(vm->constantDynamicMutex);
+			omrthread_monitor_exit(vm->constantDynamicMutex);
```
Wait should not be used bare like this - correct use is to wait for a condition to become true.

fengxue-IS commented:
Update the monitor entrance and exit around `omrthread_monitor_wait` to make sure that code is waiting on a looping check.
added fast path in resolve code to optimize performance.
fixed an error with return value not properly set.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,113 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+											J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+			if (NULL == exceptionObject) {
+				setHeapOutOfMemoryError(vmThread);
+				exceptionObject = vmThread->outOfMemoryError;
```
I believe this should be vmThread->currentException.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,113 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			resolved = true;
+		} else if (NULL != ramCPEntry->exception) {
+			J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+			if (isSameOrSuperClassOf(throwable, J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, ramCPEntry->exception))) {
```
J9VM_J9CLASS_... works on class objects - if exception is NULL, this will crash, and if it's a thread object, the results are undefined at best.  At the very least, the threadObject check needs to be first.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1847,6 +1847,113 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+				resolved = true;
+			} else if (ramCPEntry->exception == vm->voidReflectClass->classObject) {
+				/* Resolved with null reference */
+				resolved = true;
```
What happens when the BSM returns the void class object?

gacholio replied:
Nevermind, this is exception not value.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,122 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	}
+
+	if (resolved) {
+		return value;
```
Multiple returns are discouraged.  Please use if (!resolved) instead of if/return/else.

fengxue-IS replied:
fixed

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,122 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	} else {
+		/* Slow path, attempt to acquire resolve mutex or retry, or wait */
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		if (NULL == ramCPEntry->value && NULL == ramCPEntry->exception) {
```
After entering the monitor, you should do all of the checks from before the monitor enter again, like:

enter monitor
checks
if (!resolved) {
}
exit monitor
...fall through to return...

fengxue-IS replied:
inside the monitor it have a simplified block of checks that determine if there have been any update to the cpEntry and redirects with `goto retry;`

gacholio replied:
I'm not convinced it's correct, and there's no point trying to optimize the slow path instead of keeping the code the same as outside for understandability.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,129 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		ramCPEntry->exception = vmThread->threadObject;
+		omrthread_monitor_exit(vm->constantDynamicMutex);
+
+		if (!resolved) {
```
Can resolved be false at this point?

fengxue-IS replied:
yes, if this is the first time resolving or nested resolves

gacholio replied:
Right, misssed the NULL check on exception.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,129 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			}
+		}
+		ramCPEntry->exception = vmThread->threadObject;
+		omrthread_monitor_exit(vm->constantDynamicMutex);
```
Don't exit here - just do it at the end in both the resolved and unresolved paths.

fengxue-IS replied:
not sure what you meant, this is the end of all checks, anything after this point would be the actual resolution

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,129 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+				}
+			}
+		}
+		ramCPEntry->exception = vmThread->threadObject;
```
You only want to do this in the unresolved case.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,129 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		if (NULL != ramCPEntry->value) {
+			value = ramCPEntry->value;
+			resolved = true;
```
These might be better done with a retry bool instead of resolved - detect the cases that means you don't need to resolve, and go back to retry to get the value/set the exception.

fengxue-IS replied:
updated

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+					/* Another thread is currently resolving this entry */
+					// internalReleaseVMAccess(vmThread);
+					omrthread_monitor_wait(vm->constantDynamicMutex);
+					omrthread_monitor_exit(vm->constantDynamicMutex);
```
This will end up being exitting twice (below in the retry path).  Might be better to just exit and goto inline instead.

fengxue-IS replied:
I believe this exit is something I forgot to remove from last update, there are multiple retry paths, inlining all would make the code look bad. I ll remove this exit

gacholio replied:
You can't remove this one - you must exit the monitor before reacquiring VM access.

fengxue-IS replied:
Right, I see what you meant
Corrected

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+					retry = true;
+				} else if (ramCPEntry->exception != vmThread->threadObject) {
+					/* Another thread is currently resolving this entry */
+					// internalReleaseVMAccess(vmThread);
```
This is definitely required.

fengxue-IS replied:
uncommented

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+				}
+			}
+		}
+		if (retry) {
```
Given that there's one goto, may as well use all gotos and elminate the new bool (and won't have to capitalize the label).  Don't forget the monitor exit in the other two cases.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		j9object_t exceptionObject = NULL;
+		if (NULL != vmThread->currentException) {
+			/* Allocate the exception object in tenure */
+			exceptionObject = gcFuncs->j9gc_objaccess_asConstantPoolObject(
```
Dan points out that this is illegal.  The returned constant must be in the constant pool, not a tenured copy (which will not be identical).  The GC must first be updated to walk the constant pools in all policies, perhaps with an optimization tagging classes as possibly containing untenured CP objects.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		resolved = true;
+	} else {
+		/* check if already resolved to NULL or exception */
+		if (NULL != ramCPEntry->exception) {
```
Please cache this in a local rather than reading it multiple times.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+
+RETRY:
+	/* Fast path resolution for previously resolved entries */
+	if (NULL != ramCPEntry->value) {
```
This is the expected case, so could read into value, and NULL check the local.  Also means value could be declared and initialized directly here.

gacholio commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			retry = true;
+		} else {
+			/* check if already resolved to NULL or exception */
+			if (NULL != ramCPEntry->exception) {
```
Same multiple read comment as above.

fengxue-IS commented:
Updated the resolve check to use inline goto and local variables for value and exception, pending gc allocation until discussed with GC team

fengxue-IS commented:
fixed a typo in the code, updated resolved value/exception storing to use the original result object returned from the bootstrap method. the GC constant pool slot iterator will be updated in a separate PR

fengxue-IS commented:
@DanHeidinga can I get a re-review on this? The GC constant pool iterator change will be addressed in  #2457

DanHeidinga commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -252,6 +252,10 @@
 #define J9DescriptionImmediate 0x1
 
+/* Rom CP Flag for constant dynamic with primitive type */
+#define J9DescriptionCpPrimitiveType 1 << 20
```
Please add brackets to this.  Also, why 20?

fengxue-IS replied:
the slot is formatted as
12 bit unused | 16 bit bsmIndex | 4 bit Cp descriptor
20 is the least significant bit unused

DanHeidinga replied:
Good info to put into the code

DanHeidinga commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -252,6 +252,10 @@
 
+/* Rom CP Flag for constant dynamic with primitive type */
+#define J9DescriptionCpPrimitiveType 1 << 20
+#define J9DescriptionCpBsmIndexMask 0xFFFF
```
Is this constant correct if the top bit is used for primitives?

fengxue-IS replied:
this is mask is suppose to get rid of the primitive bit.
`bsmIndex = (bsmIndexAndCpType >> J9DescriptionCpTypeShift) & J9DescriptionCpBsmIndexMask`
so we only get the U_16 value for the bsmIndex

DanHeidinga commented on runtime/oti/vmconstantpool.xml:
```diff
@@ -339,6 +339,7 @@
 	<staticmethodref class="java/lang/invoke/MethodType" name="vmResolveFromMethodDescriptorString" signature="(Ljava/lang/String;Ljava/lang/ClassLoader;Ljava/lang/Class;)Ljava/lang/invoke/MethodType;" flags="opt_methodHandle"/>
 	<staticmethodref class="java/lang/invoke/MethodHandle" name="sendResolveMethodHandle" signature="(ILjava/lang/Class;Ljava/lang/Class;Ljava/lang/String;Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/invoke/MethodHandle;" flags="opt_methodHandle"/> 
 	<staticmethodref class="java/lang/invoke/MethodHandle" name="resolveInvokeDynamic" signature="(JLjava/lang/String;Ljava/lang/String;J)Ljava/lang/invoke/MethodHandle;" flags="opt_methodHandle"/> 
+	<staticmethodref class="java/lang/invoke/MethodHandle" name="resolveConstantDynamic" signature="(JLjava/lang/String;Ljava/lang/String;J)Ljava/lang/Object;" flag="opt_methodHandle"/>
```
That should include a JDK11 or similar reference as this function doesn't apply to JDK 10 & under.

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4568,11 +4568,39 @@ done:;
-		((U_32*)_sp)[1] = romConstant->slot2;
+
+		if (J9CPTYPE_CONSTANT_DYNAMIC == J9_CP_TYPE(J9ROMCLASS_CPSHAPEDESCRIPTION(J9_CLASS_FROM_METHOD(_literals)->romClass), index)) {
+			J9ConstantPool *ramConstantPool = J9_CP_FROM_METHOD(_literals);
```
Move this outside the if so it's not fetched twice.

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4568,11 +4568,39 @@ done:;
+			_sp -= 2;
+			U_64 value = J9VMJAVALANGLONG_VALUE(_currentThread, resolvedValue);
+			*(U_64*)_sp = value;
+
```
unnecessary blank line

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4568,11 +4568,39 @@ done:;
+		
+				VMStructHasBeenUpdated(REGISTER_ARGS);
+				restoreGenericSpecialStackFrame(REGISTER_ARGS);
+
```
Please add the immediate async check - see other resolves for an example

fengxue-IS replied:
async check added

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4568,11 +4568,39 @@ done:;
+				restoreGenericSpecialStackFrame(REGISTER_ARGS);
+
+				if (VM_VMHelpers::exceptionPending(_currentThread)) {
+					return GOTO_THROW_CURRENT_EXCEPTION;
```
Only return from a single point

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -6032,6 +6060,12 @@ done:;
 				break;
+			case J9DescriptionCpTypeConstantDynamic:
+				resolveConstantDynamic(_currentThread, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE);
+				if (((J9RAMConstantDynamicRef*)ramCPEntry)->exception == _vm->voidReflectClass->classObject) {
```
This needs a comment to describe why it's correct

DanHeidinga replied:
We should do the Void.class check first and let the unresolved case go around the retry loop again.
Saves a call to `resolveConstantDynamic` on every condy of null

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -6047,10 +6081,18 @@ done:;
 			value = J9VM_J9CLASS_TO_HEAPCLASS((J9Class*)value);
+		} else if ((J9DescriptionCpTypeConstantDynamic == romCPEntry->cpType & J9DescriptionCpTypeMask)
+			&& (J9_ARE_ALL_BITS_SET(romCPEntry->cpType, J9DescriptionCpPrimitiveType))
+		){
```
nitpick `){` -> `) {`

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -6047,10 +6081,18 @@ done:;
+			/* Constant Dynamic ROM CP entry uses J9DescriptionCpPrimitiveType flag to
+			 * indicate primitive return type which requires unboxing before returning the value
+			 */
+			value = (j9object_t)(UDATA)J9VMJAVALANGINTEGER_VALUE(_currentThread, value);
```
What about the other java.lang.* primitive classes like Char, Short, etc?  The shapes could be different.

DanHeidinga commented on runtime/vm/callin.cpp:
```diff
@@ -941,6 +941,53 @@ sendForGenericInvoke(J9VMThread *currentThread, j9object_t methodHandle, j9objec
+void JNICALL
+sendResolveConstantDynamic(J9VMThread *currentThread, J9ConstantPool *ramCP, UDATA cpIndex, J9ROMNameAndSignature *nameAndSig, U_16 *bsmData)
+{
+	Trc_VM_sendResolveConstantDynamic_Entry(currentThread);
```
please include the ramCP, cpIndex, nas & bsmData in the entry tracepoint

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	if (NULL != value) {
+		resolved = true;
+	} else {
+		cpException = ramCPEntry->exception;
```
move the declaration of cpException into the two scopes that use it.

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+				}
+			}
+		}
+
```
add comment to indicate this thread is the one doing the resolve

fengxue-IS replied:
comment added

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
+		U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
+		U_16 *bsmData = bsmIndices + romClass->callSiteCount;
+		U_32 bsmIndex = (romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift) & J9DescriptionCpBsmIndexMask;
```
bsmIndex needs to take the `J9DescriptionCpPrimitiveType` into account

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		U_32 bsmIndex = (romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift) & J9DescriptionCpBsmIndexMask;
+		J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+		/* Walk bsmData */
```
update the comment to describe this in more detail

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+
+		/* Walk bsmData */
+		for (U_32 i = 0; i < bsmIndex; i++) {
+			bsmData += bsmData[1] + 2;
```
Please add brackets around the RHS

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		j9object_t exceptionObject = NULL;
+		if (NULL != vmThread->currentException) {
+			/* Allocate the exception object in tenure */
+			exceptionObject = vmThread->currentException;
```
The comment and code are inconsistent.

DanHeidinga commented on runtime/bcutil/ConstantPoolMap.cpp:
```diff
@@ -503,7 +503,20 @@ ConstantPoolMap::constantPoolDo(ConstantPoolVisitor *visitor)
+				 */
+				{
+					char fieldDescriptor = (char)*_classFileOracle->getUTF8Data(getCPSlot2(cfrCPIndex));
+					if (fieldDescriptor == 'B' || fieldDescriptor == 'C' || fieldDescriptor == 'F' ||
```
Can this be written as a switch statement?  I think it would be clearer than the repeated ifs

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -828,6 +849,154 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+				classObject = JITHELPERS.getClassFromJ9Class64(j9class);
+			}
+
+			Objects.requireNonNull(classObject);
```
Please add a message to this throwable to indicate the j9class, name, etc so that issues can more easily be debugged

fengxue-IS replied:
added ExternalMessage in the catch block

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -828,6 +849,154 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+			result = MethodHandles.identity(typeClass).invoke(result);
+		} catch(Throwable e) {
+
+			/*[IF Sidecar19-SE]*/
```
none of this code is Java 9 specific - condy is 11+

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -828,6 +849,154 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+					cpEntry = getCPMethodHandleAt(internalRamClass, index);
+					break;
+				case 17:
+					// TODO - cpEntry = getCPConstantDynamicAt(internalRamClass, index);
```
please throw an InternalError for this case if we're not ready to handle it

fengxue-IS replied:
updated

fengxue-IS commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -788,6 +789,26 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+
+		if (classList.isEmpty()) {
+			throw new NoClassDefFoundError(fieldDescriptor);
+		}
```
since parseIntoClass will only parse one valid descriptor and append that to the list or throw an exception, I don't think this check is even needed.

I'll remove the check

fengxue-IS commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -925,6 +1094,9 @@ private static final MethodHandle resolveInvokeDynamic(long j9class, String name
 					cpEntry = getCPMethodHandleAt(internalRamClass, index);
 					break;
+				case 17:
+					// TODO - cpEntry = getCPConstantDynamicAt(internalRamClass, index);
```
added

DanHeidinga commented on runtime/bcutil/ConstantPoolMap.cpp:
```diff
@@ -508,13 +508,35 @@ ConstantPoolMap::constantPoolDo(ConstantPoolVisitor *visitor)
-						visitor->visitConstantDynamic(slot1, slot2, 0);
+					char fieldDescriptor = (char)*_classFileOracle->getUTF8Data(getCPSlot2(slot2));
+					switch (fieldDescriptor) {
+						case 'B':
```
nitpick: don't indent the case labels.  Write as:
```c
switch(fieldDescriptor) {
case 'B':
      visitor-......
}
This saves over indenting the code

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4591,25 +4593,37 @@ done:;
+			 * different primitive return type that require unboxing before returning the value
+			 */
+			switch (((J9ROMConstantDynamicRef *)romConstant)->bsmIndexAndCpType >> J9DescriptionReturnTypeShift) {
+				case J9DescriptionReturnTypeDouble:
```
same nitpick about indenting case labels.

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -6094,13 +6110,30 @@ done:;
 			 */
-			value = (j9object_t)(UDATA)J9VMJAVALANGINTEGER_VALUE(_currentThread, value);
+			switch (romCPEntry->cpType >> J9DescriptionReturnTypeShift) {
+				case J9DescriptionReturnTypeBoolean:
```
indenting

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1998,7 +2000,7 @@ resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 		j9object_t exceptionObject = NULL;
 		if (NULL != vmThread->currentException) {
-			/* Allocate the exception object in tenure */
+			/* Set exceptionObject to current exception */
```
nitpick: comment says exactly what the code does.  It can be deleted.

fengxue-IS commented:
fixed indentation and code comment.

DanHeidinga commented:
> Can you squash this and then I'll run the builds on it?

Never mind - the 3 commits are already reasonably organized.

DanHeidinga commented:
Jenkins test sanity xlinux jdk8

DanHeidinga commented:
Jenkins test sanity xlinux jdk11

DanHeidinga commented:
Jenkins test sanity xlinux jdk11
Jenkins test sanity xlinux jdk8

DanHeidinga commented:
Relaunching the test builds to CI issues over the weekend

DanHeidinga commented:
Jenkins test sanity xlinux jdk11

gacholio commented:
Extended testing is failing for silly reasons, so ignoring.

DanHeidinga commented:
Cancelling the JDK11 build as it fails due to missing test natives.  Merging as is as the JDK8 build passes and this will unblock more of the JDK11 work
